elastic/beats
# Infer types in Prometheus remote_write

This PR add types for Prometheus metrics that are collected via `remote_write` metricset.
These metrics are in raw format when received so a heuristic approach is applied to extract types from metrics' name patterns. 

Closes https://github.com/elastic/beats/issues/17675.

### TODOs

### How to test this PR

#### Env setup 

Use https://github.com/ChrsMark/docker-prometheus-playground for a ready to use prometheus env. 
For remote write this config is needed: https://github.com/ChrsMark/docker-prometheus-playground/blob/master/prometheus/prometheus.yml#L10
If using a `darwin` machine the following should be enough:
```
remote_write:
- url: "http://host.docker.internal:9201/write"
```
On `linux` machines in order to access host from within the Prometheus container one can set `network_mode: host` in the `docker-compose.yml` and then host can be reach using `0.0.0.0`.

#### Testing
1. Use oss version and test `collector` and `remote_write` (without `types`) metricsets for regressions.
2. Use basic version and test `collector` (with `types`) for regressions.
3. Use basic version and test that `remote_write` metricset: 
    a. uses types as expected.
    b. can identify types based on input patterns: https://github.com/elastic/beats/pull/19944/files?file-filters%5B%5D=.asciidoc#diff-b69e84fad87eb2ae563115f43675a547R118
